File: app/src/main/java/eu/kanade/tachiyomi/data/sync/SyncManager.kt
package eu.kanade.tachiyomi.data.sync

import android.content.Context
import androidx.core.net.toUri
import eu.kanade.domain.sync.SyncPreferences
import eu.kanade.domain.sync.SyncService
import eu.kanade.tachiyomi.data.backup.BackupNotifier
import eu.kanade.tachiyomi.data.backup.create.BackupCreator
import eu.kanade.tachiyomi.data.backup.create.BackupOptions
import eu.kanade.tachiyomi.data.backup.restore.BackupRestorer
import eu.kanade.tachiyomi.data.backup.restore.RestoreOptions
import eu.kanade.tachiyomi.network.NetworkHelper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import logcat.LogPriority
import tachiyomi.core.common.util.system.logcat
import uy.kohesive.injekt.Injekt
import uy.kohesive.injekt.api.get
import java.io.File
import java.time.Instant
import java.util.UUID

class SyncManager(
    private val context: Context,
    private val syncPreferences: SyncPreferences = Injekt.get(),
    private val networkHelper: NetworkHelper = Injekt.get(),
) {

    /**
     * Performs a full bidirectional sync. Two flavours depending on transport:
     *
     * - Server-mediated (SyncYomi): snapshot local → POST to server → server merges → apply merged.
     * - File-storage (WebDAV / Drive / Dropbox): pull remote payload (if any) → apply it locally
     *   (BackupRestorer does timestamp-based per-row merge) → snapshot the now-merged state →
     *   push that snapshot back to remote.
     *
     * Uploads stream straight off disk so peak memory is bounded by the merged-payload pull
     * (typically much smaller than a full library), not the full local backup size.
     *
     * Throws [SyncException] on any failure so callers can surface a useful message.
     */
    suspend fun sync() = withContext(Dispatchers.IO) {
        val service = SyncService.fromInt(syncPreferences.service().get())
        if (service == SyncService.NONE) {
            throw SyncException("Sync service not configured")
        }
        val deviceId = ensureDeviceId()
        val transport = buildTransport(service)

        val pushFile = File(context.cacheDir, "sync_push.tachibk")
        val pullFile = File(context.cacheDir, "sync_pull.tachibk")

        try {
            when (transport) {
                is SyncTransport.ServerMediated -> serverMediatedSync(transport, deviceId, pushFile, pullFile)
                is SyncTransport.FileStorage -> fileStorageSync(transport, pushFile, pullFile)
            }
        } catch (e: SyncException) {
            throw e
        } catch (e: Exception) {
            logcat(LogPriority.ERROR, e) { "Sync failed" }
            throw SyncException(e.message ?: "Sync failed", e)
        } finally {
            runCatching { pushFile.delete() }
            runCatching { pullFile.delete() }
        }
    }

    private suspend fun serverMediatedSync(
        transport: SyncTransport.ServerMediated,
        deviceId: String,
        pushFile: File,
        pullFile: File,
    ) {
        snapshotLocal(pushFile)
        val merged = transport.exchange(
            local = pushFile,
            lastSyncTimestamp = syncPreferences.lastSyncTimestamp().get(),
            deviceId = deviceId,
        )
        pullFile.writeBytes(merged)
        applyMerged(pullFile)
        syncPreferences.lastSyncTimestamp().set(Instant.now().toEpochMilli())
    }

    private suspend fun fileStorageSync(
        transport: SyncTransport.FileStorage,
        pushFile: File,
        pullFile: File,
    ) {
        // 1. Pull remote payload (if any) and apply it locally. BackupRestorer's per-row
        //    timestamp merge resolves conflicts between the remote snapshot and local state.
        val remoteBytes = transport.pull()
        if (remoteBytes != null) {
            pullFile.writeBytes(remoteBytes)
            applyMerged(pullFile)
        }

        // 2. Snapshot the now-merged local state and push it back as the new authoritative copy.
        snapshotLocal(pushFile)
        transport.push(pushFile)

        syncPreferences.lastSyncTimestamp().set(Instant.now().toEpochMilli())
    }

    private suspend fun snapshotLocal(pushFile: File) {
        pushFile.parentFile?.mkdirs()
        if (!pushFile.exists()) pushFile.createNewFile()
        BackupCreator(context, isAutoBackup = false).backup(pushFile.toUri(), buildBackupOptions())
    }

    private suspend fun applyMerged(pullFile: File) {
        val notifier = BackupNotifier(context)
        BackupRestorer(context, notifier, isSync = true).restore(
            uri = pullFile.toUri(),
            options = RestoreOptions(
                libraryEntries = true,
                categories = syncPreferences.syncCategories().get(),
                appSettings = false,
                sourceSettings = false,
                extensionRepoSettings = false,
            ),
        )
    }

    private fun buildTransport(service: SyncService): SyncTransport {
        val host = syncPreferences.host().get().trim().trimEnd('/')
        val apiKey = syncPreferences.apiKey().get()
        return when (service) {
            SyncService.NONE -> throw SyncException("Sync service not configured")
            SyncService.SYNCYOMI -> {
                if (host.isEmpty()) throw SyncException("Sync host is empty")
                SyncYomiTransport(host, apiKey, networkHelper)
            }
            SyncService.WEBDAV -> {
                if (host.isEmpty()) throw SyncException("WebDAV URL is empty")
                val username = syncPreferences.username().get()
                if (username.isBlank()) throw SyncException("WebDAV username is empty")
                if (apiKey.isBlank()) throw SyncException("WebDAV password is empty")
                WebDavTransport(host, username, apiKey, networkHelper)
            }
            SyncService.GOOGLE_DRIVE -> {
                if (apiKey.isBlank()) throw SyncException("Google Drive access token is empty")
                GoogleDriveTransport(apiKey, networkHelper)
            }
            SyncService.DROPBOX -> {
                if (apiKey.isBlank()) throw SyncException("Dropbox access token is empty")
                DropboxTransport(apiKey, networkHelper)
            }
        }
    }

    private fun buildBackupOptions(): BackupOptions = BackupOptions(
        libraryEntries = true,
        categories = syncPreferences.syncCategories().get(),
        chapters = syncPreferences.syncChapters().get(),
        tracking = syncPreferences.syncTracking().get(),
        history = syncPreferences.syncHistory().get(),
        appSettings = false,
        extensionRepoSettings = false,
        sourceSettings = false,
        privateSettings = false,
        readEntries = false,
    )

    private fun ensureDeviceId(): String {
        val existing = syncPreferences.deviceId().get()
        if (existing.isNotBlank()) return existing
        val generated = UUID.randomUUID().toString()
        syncPreferences.deviceId().set(generated)
        return generated
    }
}

class SyncException(message: String, cause: Throwable? = null) : Exception(message, cause)

File: flutter/lib/data/sync/sync_manager.dart
/// Top-level orchestration: snapshot local state → push/exchange → apply
/// merged result → record last-sync timestamp. Direct port of Mihon's
/// `SyncManager`.
///
/// Two flow shapes:
///
/// 1. **Server-mediated** (SyncYomi): build the local backup → send to
///    the server → server merges authoritatively → apply the returned
///    payload locally.
///
/// 2. **File storage** (WebDAV / Drive / Dropbox): pull whatever blob
///    exists remotely → apply it via `BackupRestorer` (per-row
///    timestamp merge) → build a fresh local backup of the now-merged
///    state → push it back as the new authoritative remote copy.
///
/// All transports throw [SyncException] on failure so callers can surface
/// a stable message in the UI.
library;

import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../data/network/app_http_client.dart';
import '../backup/backup_codec.dart';
import '../backup/backup_creator.dart';
import '../backup/backup_restorer.dart';
import '../backup/models/backup_models.dart';
import 'dropbox_transport.dart';
import 'google_drive_transport.dart';
import 'sync_preferences.dart';
import 'sync_transport.dart';
import 'syncyomi_transport.dart';
import 'webdav_transport.dart';

class SyncManager {
  SyncManager({
    required this.preferences,
    required this.creator,
    required this.restorer,
    required this.http,
  });

  final SyncPreferences preferences;
  final BackupCreator creator;
  final BackupRestorer restorer;
  final AppHttpClient http;

  /// Runs a full bidirectional sync. Returns the number of manga entries
  /// the restore step applied so the UI can show a "synced N items"
  /// snackbar.
  Future<int> sync() async {
    final data = preferences.read();
    if (data.service == SyncService.none) {
      throw SyncException('Sync service not configured');
    }

    final deviceId = await _ensureDeviceId(data);
    final apiKey = await preferences.getApiKey();
    final transport = _buildTransport(data, apiKey);

    final applied = switch (transport) {
      SyncTransportServerMediated() =>
        await _serverMediatedSync(transport, deviceId, data),
      SyncTransportFileStorage() => await _fileStorageSync(transport, data),
    };

    await preferences.setLastSyncTimestamp(
      DateTime.now().millisecondsSinceEpoch,
    );
    return applied;
  }

  Future<int> _serverMediatedSync(
    SyncTransportServerMediated transport,
    String deviceId,
    SyncPreferencesData data,
  ) async {
    final local = await encodeBackupAsync(_select(await creator.create(), data));
    final merged = await transport.exchange(
      local: local,
      lastSyncTimestamp: data.lastSyncTimestamp,
      deviceId: deviceId,
    );
    final result =
        await restorer.restore(_select(await decodeBackupAsync(merged), data));
    return result.mangaRestored;
  }

  /// Honours the "What to sync" selection, in both directions: an entity the
  /// user switched off is neither pushed to the remote nor applied from it.
  ///
  /// The creator and restorer deliberately stay selection-blind — they exist
  /// to move whole backups, and a backup file always carries everything.
  /// This is the one place the selection is enforced.
  Backup _select(Backup backup, SyncPreferencesData data) {
    if (data.syncCategories &&
        data.syncChapters &&
        data.syncTracking &&
        data.syncHistory) {
      return backup;
    }
    return Backup(
      backupManga: [
        for (final m in backup.backupManga)
          m.withSelection(
            chapters: data.syncChapters,
            categories: data.syncCategories,
            tracking: data.syncTracking,
            history: data.syncHistory,
          ),
      ],
      // Per-manga `categories` are indices into this list — the two have to
      // be dropped together or the indices dangle.
      backupCategories:
          data.syncCategories ? backup.backupCategories : const [],
      backupSources: backup.backupSources,
      backupPreferences: backup.backupPreferences,
      backupSourcePreferences: backup.backupSourcePreferences,
      backupExtensionRepo: backup.backupExtensionRepo,
      backupMangaLinks: backup.backupMangaLinks,
    );
  }

  Future<int> _fileStorageSync(
    SyncTransportFileStorage transport,
    SyncPreferencesData data,
  ) async {
    // 1. Pull whatever the remote currently holds and apply it locally
    //    first — this lets BackupRestorer's `lastModifiedAt` merge
    //    resolve per-row conflicts.
    var appliedFromRemote = 0;
    final remote = await transport.pull();
    if (remote != null) {
      final result =
          await restorer.restore(_select(await decodeBackupAsync(remote), data));
      appliedFromRemote = result.mangaRestored;
    }

    // 2. Snapshot the now-merged local state and push it back as the new
    //    authoritative copy.
    final snapshot =
        await encodeBackupAsync(_select(await creator.create(), data));
    await transport.push(snapshot);
    return appliedFromRemote;
  }

  SyncTransport _buildTransport(SyncPreferencesData data, String apiKey) {
    final host = data.host.trimRight().replaceAll(RegExp(r'/+$'), '');
    switch (data.service) {
      case SyncService.none:
        throw SyncException('Sync service not configured');
      case SyncService.syncYomi:
        if (host.isEmpty) throw SyncException('SyncYomi host is empty');
        if (apiKey.isEmpty) throw SyncException('SyncYomi API key is empty');
        return SyncYomiTransport(host: host, apiKey: apiKey, dio: http.dio);
      case SyncService.webDav:
        if (host.isEmpty) throw SyncException('WebDAV URL is empty');
        if (data.username.isEmpty) {
          throw SyncException('WebDAV username is empty');
        }
        if (apiKey.isEmpty) throw SyncException('WebDAV password is empty');
        return WebDavTransport(
          host: host,
          username: data.username,
          password: apiKey,
          dio: http.dio,
        );
      case SyncService.googleDrive:
        if (apiKey.isEmpty) {
          throw SyncException('Google Drive access token is empty');
        }
        return GoogleDriveTransport(accessToken: apiKey, dio: http.dio);
      case SyncService.dropbox:
        if (apiKey.isEmpty) {
          throw SyncException('Dropbox access token is empty');
        }
        return DropboxTransport(accessToken: apiKey, dio: http.dio);
    }
  }

  /// Mihon writes a UUID once and reuses it across syncs so the server
  /// can distinguish devices for the same account.
  Future<String> _ensureDeviceId(SyncPreferencesData data) async {
    if (data.deviceId.isNotEmpty) return data.deviceId;
    // Cheap UUID v4-ish without pulling a new dep: random hex pieces.
    final rand = DateTime.now().microsecondsSinceEpoch.toRadixString(16) +
        DateTime.now().millisecondsSinceEpoch.toRadixString(16);
    final id = '${rand.padLeft(24, '0').substring(0, 24)}-mohyeong';
    await preferences.write(data.copyWith(deviceId: id));
    return id;
  }
}

final syncManagerProvider = FutureProvider<SyncManager>((ref) async {
  final prefs = await ref.watch(syncPreferencesProvider.future);
  return SyncManager(
    preferences: prefs,
    creator: ref.watch(backupCreatorProvider),
    restorer: ref.watch(backupRestorerProvider),
    http: ref.watch(appHttpClientProvider),
  );
});
